compilers/concrete-compiler/compiler/lib/Dialect/Concrete/Analysis/MemoryUsage.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+outcome::checked<int64_t, StringError>
+getBufferSize(mlir::MemRefType bufferType) {
+  auto shape = bufferType.getShape();
+  auto elementType = bufferType.getElementType();
+  auto elementSize = getElementTypeSize(elementType);
+  if (elementSize == -1)
+    return StringError(
+        "allocation of buffer with a non-supported element-type");
+  for (auto size : shape) {
+    if (size == mlir::ShapedType::kDynamic) {
+      log_verbose() << "warning: dynamic dimension found during computation of "
+                       "memory usage. Dynamic size will be ignored";
+    }
+  }
+  auto multiply_ignore_dyn_size = [](int64_t size_1, int64_t size_2) {
+    // we don't want to multiply by a dynamic size
+    return size_2 == mlir::ShapedType::kDynamic ? size_1 : size_1 * size_2;
+  };
+
+  return elementSize * std::accumulate(shape.begin(), shape.end(), 1,
+                                       multiply_ignore_dyn_size);
+}
+
+bool isBufferDeallocated(mlir::Value buffer) {
+  for (auto user : buffer.getUsers()) {
+    if (mlir::isa<memref::DeallocOp>(user))
+      return true;
+  }
+  return false;
+}
+
+int64_t numberOfUsersInLocation(mlir::Value value, mlir::Location loc) {
+  int64_t count = 0;
+  for (auto user : value.getUsers())
+    if (user->getLoc() == loc)
+      count++;
+  return count;
+}
+
+static std::optional<StringError> on_enter(memref::AllocOp &op,
+                                           MemoryUsagePass &pass) {
+
+  auto maybeBufferSize = getBufferSize(op.getResult().getType());
+  if (!maybeBufferSize) {
+    return maybeBufferSize.error();
+  }
+  // if the allocated buffer is being deallocated then count it as one.
+  // Otherwise (and there must be a problem) multiply it by the number of
+  // iterations
+  int64_t numberOfAlloc =
+      isBufferDeallocated(op.getResult()) ? 1 : pass.iterations;
+
+  auto location = locationString(op.getLoc());
+  // pass.iterations number of allocation of size: shape_1 * ... * shape_n *
+  // element_size
+  auto memoryUsage = numberOfAlloc * maybeBufferSize.value();
+
+  pass.feedback.memoryUsagePerLoc[location] += memoryUsage;
+
+  return std::nullopt;
+}
+
+static std::optional<StringError> on_enter(mlir::Operation *op,
+                                           MemoryUsagePass &pass) {
+  for (auto operand : op->getOperands()) {
+    // we only consider buffers
+    if (!mlir::isa<mlir::MemRefType>(operand.getType()))
+      continue;
+    // allocation inside the current location are computed separately
+    auto definingOp = operand.getDefiningOp();
+    if (definingOp && mlir::isa<memref::AllocaOp>(definingOp) &&
+        definingOp->getLoc() == op->getLoc())
+      continue;
+
+    auto maybeBufferSize =
+        getBufferSize(operand.getType().cast<mlir::MemRefType>());
+    if (!maybeBufferSize) {
+      return maybeBufferSize.error();
+    }
+    auto bufferSize = maybeBufferSize.value();
+    auto numberOfUsers = numberOfUsersInLocation(operand, op->getLoc());
```

Comment:
Its a bit hacky here and I would say it's not working (more than the rounding that could over estimate a bit the memory usage).

Let's consider this example :

```
%alloc_a = memref.alloc() : memref<...>
...

%alloc_b = memref.alloc() : memref<...>
scf.for ... {
  ...
  scf.for {
    %sub_a = memref.subview %alloc_a
    %sub_b = memref.subview %alloc_b

    op(%sub_a, %sub_b)
  }
}
```

Here the `b` location has as memory usage = `%alloc_a + %alloc_b + %sub_a + %sub_b` while its actually memory footprint should be just %alloc_a + %alloc_b. I think for each location you need to have a set of the actual buffers used, and for each operand of operations you should find back the actual origin of the buffer (i.e. a block argument or the allocation) and update the set.

Typically for the location b here you will have b => %alloc_a, %alloc_b